### ISSUE TITLE
Only Allow Launch Approved Workflows to Appear in Mobile SwipeClassifier

### ISSUE THREAD

**Issue by wgranger** [zooniverse/Panoptes#2432]
In merging [this](https://github.com/zooniverse/Panoptes-Front-End/pull/3894#issuecomment-316961465) PR there is a concern that the button to add a workflow to the mobile classifier could be enabled by DOM hacking. To keep this from happening, it would be good to have a check in the backend that a workflow's `configuration.swipe_enabled` flag can only be set to `true` if the project is launch approved. Specifically, the call to toggle that flag is made in [this](https://github.com/zooniverse/Panoptes-Front-End/pull/3894/files#diff-9d06b30bfa6999a9521d6e986c1e3feeR98) function of the PR.

**Comment by camallen** [comment]
We can't control the workflow's arbitrary configuration json data. If you want this field to be used like this we should promote it out of the configuration json object to it's own attribute. Isn't this what the mobile_friendly column on a workflow was for?

**Comment by wgranger** [comment]
Ah yes. Good point. If that is the case, attaching the same check to the `mobile_friendly` tag would be sufficient. That is, in order for a project to be set `mobile_friendly: true` it would first have to be a launch_approved project. 

Consider the following:
- The mobile app only displays `mobile_friendly` projects.
- On the PR above, when a workflow is toggled `swipe_enabled: true` the corresponding project's `mobile_friendly` tag is also switched to `true` if originally set to `false`.
- However, if a check is made in the backend that a project is `launch_approved` before toggling the project to `mobile_friendly`, the workflow would not be in danger of showing up in the mobile app.

**Comment by camallen** [comment]
These states are reminding me of the terrible project states, what are we trying to achieve? 
1. Mobile friendly flag on projects & workflows
2. The ability to flag swipe enabled on workflows to wire the mobile app UI

It seems we are using the `project#mobile_friendly` but not the workflow one:
```
> Project.where(mobile_friendly: true).count
=> 15
> Workflow.where(mobile_friendly: true).count
=> 0
```
Should we just rename the workflow attribute to be `swipe_enabled` and copy the `configuration.swipe_enabled` values across?

**Comment by wgranger** [comment]
Interesting, I was unaware there was also a mobile_friendly attribute attached to the workflow. Yes, it is definitely redundant to have `workflow.mobile_friendly` `workflow.configuration.swipe_enabled`, so removing the configuration values over would make sense. I'll have to adjust the mobile code a bit so workflows look for the mobile_friendly tag when opening the swipe classifier.

I believe we initially created both the `mobile_friendly` and `configuration.swipe_enabled` tags because we thought there might be a mobile friendly project that could have a complicated workflow that shouldn't open on mobile. However, any workflow that isn't swipe enabled will open in a webview regardless.

**Comment by camallen** [comment]
> I'll have to adjust the mobile code a bit so workflows look for the mobile_friendly tag when opening the swipe classifier.

We can copy the `configuration.swipe_enabled` value over and leave it as is to support older versions of the app with new versions using the workflow setting. 

Should we convert the `workflow.mobile_friendly` to `workflow.swipe_enabled` ? will it only be used to activate the swipe UI?

> However, any workflow that isn't swipe enabled will open in a webview regardless.

Agree that the webview fixes most of this - especially if the classifier UX upgrade makes it render better on smaller screens.

**Comment by wgranger** [comment]
This leads to a more long term convo of what we expect from the mobile app. Currently, the **only** workflows that open within the mobile app are with the SwipeClassifier, so things are a bit simple now.

However, with the hiring of a mobile dev down the line, what else do we expect from the mobile app? If we plan to open other workflows within the app (not just the swipe classifier) we may want to denote that a workflow is `mobile_friendly` to open within the app and not just webviews. 

`swipe_enabled` is definitely more specific for this use case and how the app is currently operating as having both `workflow.mobile_friendly` and `project.mobile_friendly` is quite confusing.

**Comment by camallen** [comment]
@wgranger what happens in the app if the workflow `configuration.swipe_enabled` via hacking? I assume the project doesn't load initially and thus won't do anything to activate that project in the app?

I think the best way forward is change the `workflow.mobile_friendly` attribute to be `worklfow.swipe_enabled` and ensure the former and the `project.mobile_friendly` attribute can only be set by admins.

**Comment by wgranger** [comment]
A project has to be tagged `mobile_friendly` to even appear in an app since the mobile app makes an api call upon load for all mobile projects. Even if a workflow is tagged `swipe_enabled` it won't appear in the app unless the project has a mobile tag.

Yes, I think `workflow.mobile_friendly` should be changed to `workflow.swipe_enabled`. 

I'm not sure about only allowing admins to set that project tag though. The purpose of the original [PR](https://github.com/zooniverse/Panoptes-Front-End/pull/3894) is to give project owners the ability to tag a workflow swipe enabled in the project builder, and that toggle will also activate `project.mobile_friendly` if it is otherwise turned off. I think only restricting this to `launch_approved` projects should be sufficient since we would've screened the project.

If we only allow admins to set the `project.mobile_friendly` tag, it seems to defeat the purpose of the PR above since a project owner will still be reliant on a Zooniverse member to toggle the `project.mobile_friendly` tag.

**Comment by camallen** [comment]
> A project has to be tagged mobile_friendly to even appear in an app since the mobile app makes an api call upon load for all mobile projects. Even if a workflow is tagged swipe_enabled it won't appear in the app unless the project has a mobile tag.

Excellent - just to be clear it won't show in the app unless the project is mobile_friendly AND launch approved?

> I'm not sure about only allowing admins to set that project tag though. The purpose of the original PR is to give project owners the ability to tag a workflow swipe enabled in the project builder, and that toggle will also activate project.mobile_friendly if it is otherwise turned off. I think only restricting this to launch_approved projects should be sufficient since we would've screened the project.

Understood, is there any way we can check they've setup the workflow for swipe correctly to make sure the app doesn't break? @mrniaboc thoughts on allowing project owners of launch approved projects to turn on swipe enabled workflows without us checking them?

**Comment by marten** [comment]
As a project owner, how would I test my workflow in the app before making it public / marking it as mobile-friendly?

**Comment by mrniaboc** [comment]
I really don't see an issue with allowing the project owner to set a workflow to swipe enabled, assuming it can only be enabled for binary options and there is clear text describing the feature ad how it works.

**Comment by wgranger** [comment]
A project will appear in the app if it is tagged mobile_friendly. Of those projects, a workflow will only open in the app if tagged swipe_friendly (what we're changing the tag to). All other workflows open in a webview. Currently, workflows tagged `configuration.swipe_enabled` are the only ones that open in the app.

There are two checks to make sure only an appropriate workflow will open in swipe classifier. One check is in the project builder (where a project owner will toggle a workflow for the swipe classifier) another check is made in the app itself (to make sure a workflow is appropriate before attempting to open it).

**Comment by wgranger** [comment]
@marten Currently, there is not an environment set up where a project owner can test a workflow before toggling it. This is something we should aim for once a mobile dev joins our team.

**Comment by wgranger** [comment]
I'm reopening this since we're still waiting on this backend change to merge a frontend addition.

**Comment by camallen** [comment]
I'm still not sure what i'm meant to do with this. can you summarize the required changes you want from the comments above?

**Comment by wgranger** [comment]
We don't need the `workflow.mobile_friendly` tag anymore. Instead, we now need the `workflow.swipe_enabled` tag. 

Also `workflow.swipe_enabled` can only be toggled to `true` if the project associated with that workflow has `project.mobile_friendly` set to true.

**Comment by marten** [comment]
I think `swipe_enabled` is be too specific. It's not like it's a swipe app, it's a mobile app. What if add support for mobile surveys?

**Comment by wgranger** [comment]
After having a conversation with @marten, we decided it would be just as effective to adjust the project query in mobile to `launch_approved: true` without making any backend changes. 

This way, even if a project owner DOM hacks the project builder to toggle a workflow `mobile_friendly`, that project still won't be shown in the App.
